### ISSUE TITLE
Keybindings for directionnal terminal switch

### ISSUE THREAD

**Issue by noirbizarre** [gnunn1/terminix#136]
Hi !

Previous and next terminal keybindings are sufficiant when you split only once but when you have, let's say 4 terms, it become difficult to go to the expected term.
It would be nice to have shortcut for:

- go to the left terminal
- go to the right terminal
- go to the above terminal
- go to the terminal below

**Comment by gnunn1** [comment]
It's an interesting idea but no plans to do this, because of the way the Paned component component figuring out which is the right terminal based on direction is fairly complicated and I suspect error-prone.

**Comment by alexwhitman** [comment]
Just for reference it's implemented in terminator [here](http://bazaar.launchpad.net/~gnome-terminator/terminator/trunk/view/head:/terminatorlib/window.py#L804)

It would be useful to have, I use it quite a bit in terminator.  Maybe it could be something for a plugin if plugins are on the roadmap?

**Comment by gnunn1** [comment]
@alexwhitman Thanks, that's pretty much along the lines of what I'd expect the implementation to look like. I'll keep this feature request in mind in case I have a eureka moment but it still likes kind of error prone to me. I don't use Terminator so no experience with this feature, do you find it consistently makes the expected choice, i.e. no surprises in what terminal it selects for a direction?

No plans for a plugin capability at this time, since Terminix is open source and people can just submit pull requests for now.

**Comment by alexwhitman** [comment]
> do you find it consistently makes the expected choice, i.e. no surprises in what terminal it selects for a direction?

That tends to depend on the layout but I think it can be improved.  For example, if you vertically split once, then in the right hand side horizontally split so you have:
```
|-----|-----|
|     |  2  |
|  1  |-----|
|     |  3  |
|-----|-----|
```
Moving right from 1 ends focuses 3 but I would personally expect it to focus 2.

I've never used D before but if I get some time I might take a look at implementing it.

**Comment by phw** [comment]
That's actually the one feature I miss most in Terminix which prevents me to completely abandon Terminator. I find myself using this way of navigating between the terminals all the time.

**Comment by gnunn1** [comment]
@alexwhitman From a design perspective, I would think a simple brute force approach would actually work best over trying to do anything to fancy. Using your example layout, I'd look at driving the direction off of the caret position in the focused terminal and then start sampling every 20-40 pixels or so in the desired direction until you hit a different terminal or are out of screen. Using that approach if the caret is at the top of 1, 2 would be selected. If the caret was at the bottom 3 would be selected.

If you want to work on this and submit a pull request I'd be happy to take it. Some considerations for you if you do tackle it:

a. You want to make your code changes in session.d, that's where the management of terminals and panes is located. You will need a public method that asks to focus a terminal based on direction, see the existing focusNext, focusPrevious methods
b. Make sure to use the focusTerminal method in session.d to actually do the focusing
c. You can add new session actions in appwindow.d in the method createSessionActions, see the focusNext and focusPrevious actions there
d. Update the file com.gexperts.Terminix.gschema.xml to add new keybindings for the new actions

I'm going to re-open this issue since there seems to be enough demand and remove the won't fix tag, if you do decide to work on this, just post a note mentioning it here so no one (myself most likely) doesn't step on your toes.

**Comment by alexwhitman** [comment]
@gnunn1 Thanks for the pointers.  I've made a start on this in https://github.com/alexwhitman/terminix/tree/directional-terminal-switch - so far it just hooks up the directional actions.  Next step is to actually implement the terminal switching, I'll need to get familiar with GTK.

**Comment by gnunn1** [comment]
Closing thanks to the work of @alexwhitman, great job Alex! Feel free to work on other features :)

**Comment by noirbizarre** [comment]
OMG !!! Thanks @alexwhitman  !!!!
